
# STEP 1 - FINDING VARIANTS IN MASTER FILE
Here every variant found in single patient (all cells combined) are found

__data at the start of this step: a merged bam file of all individiual bam files per cell__

## 1.1 - generate a mpileup file from the data
example usage: 
> $ python3 1_1_findReadsperPosition.py sorted_run_1_cell_1.bam hg19_withchr.fa run1_cell1_counts.txt
input 1: the merged bam file
input 2: refrence genome in fasta format
**It is import that the chromosomes in the refrence file are noted with chr I.E. chr1 not simply a 1 
input 3: the output name of the count file that is generated

In the same folder as 1_1_findReadsperPosition.py should also be, pileup2base.pl.

## 1.2 - select all variant positions
example usage: 
> $ python3 1_2_selectVariants.py run1_cell1_counts.txt run1_cell1_variants.txt
input 1: a count file generated by mpilupe and parsed
input 2: the output name of the variant txt file

## 1.3 - merge all variant files of a single patient
_In some cases the data of a patient is gathered in multiple runs. It is the easiest to merge all cells per run and perform step 1_1 and 1_2 seperately for every run.
Then merge the generate variant file per run in this step._
example usage:
> $ python3 1_3_combine_variantFiles.py variantsFiles.txt patient1_run123_variants.txt

input 1: txt file which contains the relative path to every variant file, one per line
input 2: the output name of the merged variant file
